### ISSUE TITLE
hwmonitor@sylfurd.net : Crashes and disappears after upgrade

### ISSUE THREAD

**Issue by Hultan** [linuxmint/cinnamon-spices-applets#3005]
```
 * Applet version/Build date (to be found in "About.." or in "metadata.json")
1.3.0
 * Cinnamon version (cinnamon --version)
4.4.8
 * Distribution - (Mint 17.2, Arch, Fedora 25, etc...)
Linux Mint
 * Graphics hardware *and* driver used
AMD
 * Snippet or attachment of ~/.cinnamon/glass.log
 * Snippet or attachment of ~/.xsession-errors
```
Cjs-Message: 17:10:20.880: JS LOG: [LookingGlass/info] Installing settings for hwmonitor@sylfurd
Cjs-Message: 17:10:20.880: JS LOG: [LookingGlass/info] Settings successfully installed for hwmonitor@sylfurd
Cjs-Message: 17:10:20.894: JS LOG: [LookingGlass/error] Exception in getDiskLoad(): device is null
Cjs-Message: 17:10:20.894: JS LOG: [LookingGlass/error] 
[hwmonitor@sylfurd]: this.getDiskLoad(...) is undefined
[hwmonitor@sylfurd]: Failed to evaluate 'main' function on applet: hwmonitor@sylfurd/14
Cjs-Message: 17:10:20.894: JS LOG: [LookingGlass/trace] 
<----------------
DiskDataProvider@/home/per/.local/share/cinnamon/applets/hwmonitor@sylfurd/3.8/providers.js:225:47
anonymous/GraphicalHWMonitorApplet.prototype.createAppletArea@/home/per/.local/share/cinnamon/applets/hwmonitor@sylfurd/3.8/applet.js:240:37
anonymous/GraphicalHWMonitorApplet.prototype._init@/home/per/.local/share/cinnamon/applets/hwmonitor@sylfurd/3.8/applet.js:170:9
GraphicalHWMonitorApplet@/home/per/.local/share/cinnamon/applets/hwmonitor@sylfurd/3.8/applet.js:68:5
main@/home/per/.local/share/cinnamon/applets/hwmonitor@sylfurd/3.8/applet.js:487:12
createApplet@/usr/share/cinnamon/js/ui/appletManager.js:581:18
addAppletToPanels@/usr/share/cinnamon/js/ui/appletManager.js:356:22
onEnabledAppletsChanged@/usr/share/cinnamon/js/ui/appletManager.js:296:9
---------------->

(csd-power:1660): power-plugin-CRITICAL **: 17:10:20.903: abs_to_percentage: assertion 'max > min' failed
```

 ```

**Notify author of applet**
@jacobwills 

**Issue**

When I upgraded on one of my machines the applet "disappeared", because of the crash above. It is still technically on the panel, but it is not visible.

~~Note that there is something weird going on here, the log above are referencing the 3.8-version of the applet, but it shouldn't since the cinnamon version is 4.4.8.~~ My mistake, that is correct behavour, it is version 3.2 and 3.8 we have (not 3.8 and 4.2 that I for some reason thought :-).

Here is the disk information of the disk available on that machine, in case that causes the null:

![image](https://user-images.githubusercontent.com/24316220/80993802-fe5ddb80-8e3b-11ea-878e-9693aba1c03b.png)

**Steps to reproduce**

Just upgrading...

**Expected behaviour**

Not crash...

**Other information**

I can look at this, probably this weekend, but you @jacobwills might wanna take a look at it before that.

**Comment by jacobwills** [comment]
I’ll definitely take a look at this today. I think it’s loading a list of disks from your save configuration and the selected one is no longer in the system or named the same. Can solve by maybe catching this and calling the getAvailableDisks method again to avoid, BUT I was under the impression this would get run when you load the config anyway because it forces a restart of the applet. Am I wrong in that? Does it only send the settings changed signal and continue? Even then it should get to the restartGHW() method also. Not yet yet. Thanks for bringing to my attention.

**Comment by Hultan** [comment]
Hmm...I have not changed the disks in the computer, that got the error, for a long time though? In fact, I don't think I have had the configuration dialog open on that machine in months. Could it be a problem if it is the first time it opens, and no disk  is selected?

If I remember correctly, the applets configurations is loaded automatically. At least I don't remember any code specifically telling the applet to load (or convert) the configs. I think that is handled in the background by Cinnamon, but I might be completely wrong about that.

> BUT I was under the impression this would get run when you load the config anyway because it forces a restart of the applet. Am I wrong in that?

I don't know, I started improving this applet about a year ago, and although I have learnt a few things about applets, I am **far** from an expert...it is mostly trial and error for me...

**Comment by jacobwills** [comment]
No sweat. I'm still learning about Cinnamon's API myself.

I've finally gotten around to testing this today but I'm not able to reproduce this error given the info you've provided me.

Questions about your setup
---

- Is the issue isolated to NVMe devices in your experience? (I don't have one to test. I'm pretty sure the M.2 slot in my system is SATA only)
- can you attach those error files to this thread and also include the applet user settings file located here: `~/.cinnamon/configs/hwmonitor@sylfurd/<applet-id>.json`?
- Are you able to get it working again by selecting a disk from the settings menu or is this not accessible after the crash, even through the Cinnamon Applets menu?
- What behavior do you get by adding a new instance or resetting to default settings?
- Can you manually load your saved settings from a file that was exported using the import/export dialog?

Just curious regarding the last 3 things. I know we are trying to ensure the update from one version to the next works properly without the user needing to perform those workarounds.

My testing methodology
---

Please let me know if this is flawed in some way.

- use `vagrant` to spin up a VirtualBox VM running the latest Linux Mint 19.3 (script updates base packages)
- `git clone` this repo
- `git reset` to the previous commit to get the last version
- copy `hwmonitor@sylfurd` into the applets folder
- restarted `cinnamon` (in order for it to show up in the applets list)
- add the applet through settings menu (v1.2 works as expected)
- I changed width of CPU panel in the applet settings to ensure it writes a custom user `settings-schema.json` (I left the disk setting as the default "/")
- Get the update using SpicesUpdate applet.
- Works! (does not disappear as described, displays disk IO graph, disk setting defaults to sda)

**Comment by jacobwills** [comment]
I think that this is caused by the default of `sda` in the default settings and there not being a device with that name on your system. I changed the default to something I don't have in my system and it broke it with the same message you had. I thought that was one of the cases I planned for and I don't remember this being a problem but it certainly doesn't like it.

I'm working something up to override the defaults with the first item from the list or otherwise fail more gracefully than this. I'm not sure if I can force it generate a user custom settings file without interaction, but since calling `settings.setValue()` causes the 'settings-changed` signal to be sent which restarts the applet that should be it. Not working yet though.

**Comment by Hultan** [comment]
You are absolutely right, there is no **sda** on that machine, so that is probably the problem on both my machines that I have the problem on now. I got the same problem on my laptop last night.

![image](https://user-images.githubusercontent.com/24316220/81139531-5abd1a00-8f66-11ea-81d1-3e33cf179b7e.png)

Selecting the first item in the list should solve the problem though. You can also just handle the error and return 0% usage if there is an error, log it, and then the user would have to look into the configurations to change drive.

We could work on an red error background color for the graphs that fails to calculate there usage value, signaling to the user that something is wrong with the graph settings? Maybe in addition with a tool tip describing what is wrong.

**Comment by Hultan** [comment]
I just wanted to add that, yes, there were no disk selected on the disk read or write tab in the configuration window, so almost certainly the **sda**-problem. When I selected a disk in the configuration window and restarted cinnamon (CTRL + ALT + ESCAPE) it now works fine.

**Comment by jacobwills** [comment]
@Hultan 

I submitted a PR that fixes this issue. Thanks for confirming the problem for me. I would like to add some of the user feedback functionality you referred to. I think that would be a great idea, but I think it will be better to add these in a separate PR as that would be more of a feature add. I'll be working on that over the next few days as time permits. Just an FYI, I am in the process of moving from Virginia to Florida, USA so my time might be limited to work on this but I'll do what I can. Thanks.

**Comment by BerekHalfhand** [comment]
I see that it's been fixed, but I'm getting a crash in Cinnamon 5.2.7  with logs
```
Cjs-Message: 09:14:10.157: JS LOG: [LookingGlass/error] Exception in getDiskLoad(): device is null
Cjs-Message: 09:14:10.158: JS LOG: [LookingGlass/error] Exception in getData():this.getDiskLoad() is undefined
Cjs-Message: 09:14:10.158: JS LOG: [LookingGlass/error] Exception in getDiskLoad(): device is null
Cjs-Message: 09:14:10.159: JS LOG: [LookingGlass/error] Exception in getData():this.getDiskLoad() is undefined
glibtop(c=2283): [ERROR] open (/proc/stat): Too many open files
```

Which apparently connected to hwmonitor@sylfurd Version 1.3.1 (2021-12-16 13:59:23)
This started recently, after i installed and removed gnome, and also mucked around with nvidia driver a lot. Not sure how it can be related. I can provide further info if needed.
